### PR TITLE
Use strict version of subcomponents when building the full bundle RPM

### PR DESCRIPTION
**Pull request by gaetanmaisse**
**Issue**

https://github.com/gravitee-io/issues/issues/7484

## Description

Having a `>=` is causing troubles as for instance it leads to downloading components 3.17.0 when you try to install
the 3.15.6 full bundle. 

This PR:

1. Use a strict `=` reference in the full bundle
2. Add some checks in the preinstall script to avoid any automatic and unwanted downgrade when users will update their installation with the next released version that will contain that fix.

### CentOS

Build the RPMs using` build.sh`:

```shell
cd apim/3.x
./build.sh -v 3.15.6
```

Run centos docker container with a volume mount:
```shell
export PATH_TO_LOCAL_RPMS=$(pwd)
docker run --rm -v ${PATH_TO_LOCAL_RPMS}:/local-rpms -it --entrypoint bash centos
```

Inside the container:
```shell
# Add packagecloud registry to be able to run install APIM
echo "[graviteeio]
name=graviteeio
baseurl=https://packagecloud.io/graviteeio/rpms/el/7/\$basearch
gpgcheck=0
enabled=1
gpgkey=https://packagecloud.io/graviteeio/rpms/gpgkey
sslverify=1
sslcacert=/etc/pki/tls/certs/ca-bundle.crt
metadata_expire=300" > /etc/yum.repos.d/graviteeio.repo

# Some dark magic to have everything related to repo working properly on my local docker container
cd /etc/yum.repos.d/
sed -i 's/mirrorlist/#mirrorlist/g' /etc/yum.repos.d/CentOS-*
sed -i 's|#baseurl=http://mirror.centos.org|baseurl=http://vault.centos.org|g' /etc/yum.repos.d/CentOS-*
yum install pygpgme yum-utils
yum update -y
yum -q makecache -y --disablerepo='*' --enablerepo='graviteeio'

# Install APIM 3.15.0, or in reality 3.17.1 :p
yum install graviteeio-apim-3x-3.15.0-0.noarch -y

# Try to install fixed version of the packages
yum install /local-rpms/graviteeio-apim-gateway-3x-3.15.6-0.noarch.rpm
yum install /local-rpms/graviteeio-apim-management-ui-3x-3.15.6-0.noarch.rpm
yum install /local-rpms/graviteeio-apim-portal-ui-3x-3.15.6-0.noarch.rpm
yum install /local-rpms/graviteeio-apim-rest-api-3x-3.15.6-0.noarch.rpm
```

All the 4 last lines (trying to install local packages) should fail with the following message:
image

To check everything is ok when upgrading run:

```shell
# Downgrade to 3.15.0
yum install graviteeio-apim-gateway-3x-3.15.0
# Upgrade to 3.15.6
yum install /local-rpms/graviteeio-apim-gateway-3x-3.15.6-0.noarch.rpm

yum install graviteeio-apim-management-ui-3x-3.15.0
yum install /local-rpms/graviteeio-apim-management-ui-3x-3.15.6-0.noarch.rpm

yum install graviteeio-apim-portal-ui-3x-3.15.0
yum install /local-rpms/graviteeio-apim-portal-ui-3x-3.15.6-0.noarch.rpm

yum install graviteeio-apim-rest-api-3x-3.15.0
yum install /local-rpms/graviteeio-apim-rest-api-3x-3.15.6-0.noarch.rpm
```

Finally to test a fresh install:
```shell
# Remove 
yum remove graviteeio-apim-gateway-3x
# Install local 3.15.6
yum install /local-rpms/graviteeio-apim-gateway-3x-3.15.6-0.noarch.rpm

yum remove graviteeio-apim-management-ui-3x
yum install /local-rpms/graviteeio-apim-management-ui-3x-3.15.6-0.noarch.rpm

yum remove graviteeio-apim-portal-ui-3x
yum install /local-rpms/graviteeio-apim-portal-ui-3x-3.15.6-0.noarch.rpm

yum remove graviteeio-apim-rest-api-3x
yum install /local-rpms/graviteeio-apim-rest-api-3x-3.15.6-0.noarch.rpm
```

### RedHat

To test the fix on RedHat just follow the CentOS section with a redhat ubi docker image:

```shell
docker run --rm -v ${PATH_TO_LOCAL_RPMS}:/local-rpms -it --entrypoint bash redhat/ubi8
```

### Debian

⚠️ There are some references to Debian in the code BUT there is absolutely nothing about Debian installation in docs.gravitee.io so I'm not sure what to do/test with it.

### AM

To test the AM RPMs follow the CentOS section and replace APIM package by `graviteeio-am-3x-3.15.0-0.noarch`

### CODE REVIEW

File: am/3.x/build.sh [hunk 1]
```diff
@@ -123,52 +123,52 @@ build_management_ui() {
...
+            -s dir -v ${VERSION}  \
+            --license "${LICENSE}" \
+            --vendor "${VENDOR}" \
+            --maintainer "${MAINTAINER}" \
+            --architecture ${ARCH} \
+            --url "${URL}" \
+            --description  "${DESC}" \
+            --depends "${PKGNAME}-management-ui-3x = ${VERSION}" \
+            --depends "${PKGNAME}-management-api-3x = ${VERSION}" \
+            --depends "${PKGNAME}-gateway-3x = ${VERSION}" \
```
Comment (gaetanmaisse) [inline am/3.x/build.sh:169]: Real changes are here

File: apim/3.x/build.sh [hunk 2]
```diff
@@ -186,24 +186,24 @@ build_full() {
...
+            --vendor "${VENDOR}" \
+            --maintainer "${MAINTAINER}" \
+            --architecture ${ARCH} \
+            --url "${URL}" \
+            --description  "${DESC}" \
+            --depends "${PKGNAME}-portal-ui-3x = ${VERSION}" \
+            --depends "${PKGNAME}-management-ui-3x = ${VERSION}" \
+            --depends "${PKGNAME}-rest-api-3x = ${VERSION}" \
+            --depends "${PKGNAME}-gateway-3x = ${VERSION}" \
+            --verbose \
```
Comment (gaetanmaisse) [inline apim/3.x/build.sh:205]: And here

**Review by ytvnr — APPROVED** [review]
is there a risk of breaking change for those who were not strict on the version ? 
maybe introduce a --nonStrict option or I don't know ?

**Comment by gaetanmaisse** [comment]
⚠️  DO NOT MERGE

**Comment by gaetanmaisse** [comment]
The `checkVersionUpgrade` function is duplicated in all the preinstall scripts, I would love to learn how to avoid it, if you have any idea feels free to try.

**Comment by brasseld** [comment]
> The `checkVersionUpgrade` function is duplicated in all the preinstall scripts, I would love to learn how to avoid it, if you have any idea feels free to try.

I was struggling to find a solution for this when I initially wrote (and enjoyed to do it !) these scripts, but found nothing as well...

**Review by brasseld — APPROVED** [review]
It is ok for me.

Not the sexier solution but working fine. Seems that we are limited by the FPM tools for doing things other way like with precheck / verifyscript but FPM does not support it.

**Pull request merged**